GoMeet/src/test/java/com/google/sps/GoogleCalendarTimesServletTest.java

```diff
@@ -0,0 +1,232 @@
+  private final String TIME_10AM = "2021-02-11T10:00";
+  private final String TIME_12PM = "2021-02-11T12:00";
+  private final String TIME_2PM = "2021-02-11T14:00";
+  private final String TIME_4PM = "2021-02-11T16:00";
+
+  private Date periodStart; 
+  private Date periodEnd;
+  
+  // Mocks
+  private HttpServletRequest mockedRequest;       
+  private HttpServletResponse mockedResponse;
+  private StringWriter stringWriter;
+  private PrintWriter writer;
+
+  // Spies
+  private GoogleCalendarTimesServlet servletSpy;
+
+  @Before
+  public void setUp() throws IOException, ParseException {
+    mockedRequest = mock(HttpServletRequest.class);      
+    mockedResponse = mock(HttpServletResponse.class);
+    stringWriter = new StringWriter();
+    writer = new PrintWriter(stringWriter);
+    when(mockedResponse.getWriter()).thenReturn(writer);
+
+    // Mock the return values from getting parameters of request query string
+    when(mockedRequest.getParameter(
+      GoogleCalendarTimesServlet.QueryStringFieldNames.GUEST_LIST))
+        .thenReturn(GUEST_LIST_STR);
+    when(mockedRequest.getParameter(
+      GoogleCalendarTimesServlet.QueryStringFieldNames.DURATION_MINS))
+        .thenReturn(DURATION_MINS);
+    when(mockedRequest.getParameter(
+      GoogleCalendarTimesServlet.QueryStringFieldNames.DURATION_HOURS))
+        .thenReturn(DURATION_HOURS);
+    when(mockedRequest.getParameter(
+      GoogleCalendarTimesServlet.QueryStringFieldNames.PERIOD_START))
+        .thenReturn(PERIOD_START_STR);
+    when(mockedRequest.getParameter(
+      GoogleCalendarTimesServlet.QueryStringFieldNames.PERIOD_END))
+        .thenReturn(PERIOD_END_STR);
+    
+    // Initialise the Dates here to recognise the ParseException.
+    periodStart = FORMATTER.parse(PERIOD_START_STR);
+    periodEnd = FORMATTER.parse(PERIOD_END_STR);
+
+    // Initialise a new spy every test to ensure atomicity
+    servletSpy = spy(new GoogleCalendarTimesServlet());
+  }
+  
+  // Tests for doGet
+  @Test 
+  public void testSuccessfulDoGet() throws IOException, ParseException {
+    List<TimePeriod> freeTimes = Arrays.asList(
+      timePeriodFromDatestring(TIME_10AM, TIME_12PM),
+      timePeriodFromDatestring(TIME_2PM, TIME_4PM)
+    );
+
+    doReturn(freeTimes).when(servletSpy)
+        .autoProposeTimes(mockedResponse, GUEST_LIST, periodStart, periodEnd, DURATION_MS);
+
+    servletSpy.doGet(mockedRequest, mockedResponse);
+
+    // Check that the autoProposeTimes function was actually called.
+    verify(servletSpy)
+        .autoProposeTimes(mockedResponse, GUEST_LIST, periodStart, periodEnd, DURATION_MS);
+
+    // Check that the response is the expected.
+    List<String> expectedReturn = Arrays.asList(
+        // The start times of the free time blocks.
+        "2021-02-11T10:00:00.000Z",
+        "2021-02-11T14:00:00.000Z"
+    );
+    Gson gson = new Gson();
+    writer.flush();
+    assertTrue(stringWriter.toString().contains(gson.toJson(expectedReturn))); 
```

Comment:
nit: Similar to how when testing if a hash-map is correctly sent, we convert the response to a hash-map and check values, I think it would be good to convert stringWriter.toString() to a list and check if it contains the correct strings.